Loop/Managers/LoopAppManager.swift

```diff
@@ -0,0 +1,307 @@
+//
+//  LoopAppManager.swift
+//  Loop
+//
+//  Created by Darin Krauss on 2/16/21.
+//  Copyright © 2021 LoopKit Authors. All rights reserved.
+//
+
+import UIKit
+import LoopKit
+import LoopKitUI
+
+public protocol ViewControllerProvider: AnyObject {
+    var viewController: UIViewController? { get set }
+}
+
+class LoopAppManager: NSObject {
+    private enum State: Int {
+        case initialize
+        case launchManagers
+        case launchOnboarding
+        case launchHomeScreen
+        case launchComplete
+
+        var next: State { State(rawValue: rawValue + 1) ?? .launchComplete }
+    }
+
+    private var launchOptions: [UIApplication.LaunchOptionsKey: Any]?
+    private var window: UIWindow?
+
+    private var pluginManager: PluginManager!
+    private var bluetoothStateManager: BluetoothStateManager!
+    private var alertManager: AlertManager!
+    private var loopAlertsManager: LoopAlertsManager!
+    private var trustedTimeChecker: TrustedTimeChecker!
+    private var deviceDataManager: DeviceDataManager!
+    private var onboardingManager: OnboardingManager!
+
+    private var state: State = .initialize
+
+    private let log = DiagnosticLog(category: "LoopAppManager")
+
+    // MARK: - Initialization
+
+    func initialize(with launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        precondition(state == .initialize)
+
+        self.launchOptions = launchOptions
+
+        registerBackgroundTasks()
+
+        self.state = state.next
+    }
+
+    func launch(into window: UIWindow?) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        precondition(!isLaunchComplete)
+        precondition(state != .initialize)
+
+        guard isProtectedDataAvailable() else {
+            log.default("Protected data not available; deferring launch...")
+            return
+        }
+
+        self.window = window
+        
+        window?.tintColor = .loopAccent
+        OrientationLock.deviceOrientationController = self
+        UNUserNotificationCenter.current().delegate = self
+
+        resumeLaunch()
+    }
+
+    var isLaunchComplete: Bool { state == .launchComplete }
+
+    private func resumeLaunch() {
+        launchManagers()
+        launchOnboarding()
+        launchHomeScreen()
+    }
+
+    private func launchManagers() {
+        dispatchPrecondition(condition: .onQueue(.main))
+
+        guard state == .launchManagers else {
```

Comment:
[very tiny nit] this seems like a programming error if not so?  Maybe consider `assert(state == .launchManagers)`?

Reply:
It is not a programming error. The LoopAppManager is a bit of a state machine. Initializing and launching the app is a multi-step (multi-state) process). First it performs some core initialization even before first unlock. Then, assuming after first unlock, it needs to set up all of the managers. Then, assuming onboarding hasn't yet been completed, the onboarding needs to present all of its UI/functionality. Then, once onboarding is entirely complete, the main home page (StatusTableViewController) can be built and installed.

So, each of the `launch*` functions handle a different state. This `guard` statement says "if I'm not in the state where I should be doing this, then bail".

Reply:
And, to clarify, if I used `assert` the app would die. We don't want it to die, we just want it to not do this work (because it has already been done some time previously).

Reply:
I suppose I could update `resumeLaunch` from:
```
    private func resumeLaunch() {
        launchManagers()
        launchOnboarding()
        launchHomeScreen()
    }
```
to:
```
    private func resumeLaunch() {
        if state == .launchManagers {
            launchManagers()
        }
        if state == .launchOnboarding {
            launchOnboarding()
        }
        if state == .launchHomeScreen {
            launchHomeScreen()
        }
    }
```
and then add a precondition check in each of the `launch*` functions.

Reply:
Ah, I understand now.  Yeah it would be slightly cleaner (and more DWIM) to do it the way you suggest, instead of the `guard` statements making them noops if they are not necessary.  Up to you.